Add content (#37)
nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook: guidance/01-reproducible/pages/01-documented.qmd

---
title: "Documented"
description: "Effective documentation is crucial for the health, usability, and maintainability of any software project."
---

> ## Pre-requisite reading
>
> * [AF Duck book: Code documentation](https://best-practice-and-impact.github.io/qa-of-code-guidance/code_documentation.html)
> * [AF Duck book: Project documentation](https://best-practice-and-impact.github.io/qa-of-code-guidance/project_documentation.html)
> * [NHSBSA DDaT playbook: Licensing software or code](https://nhsbsa.github.io/nhsbsa-digital-playbook/development/coding-licences/)

> > * [NHSBSA DDaT playbook: Standard code repository files](https://nhsbsa.github.io/nhsbsa-digital-playbook/development/dev-documentation/)

Author: Done

> Effective documentation is crucial for the health, usability, and maintainability of any software project. It encompasses both the specifics within the code and the broader context of the project itself.

Author: Took a few tries to get that second sentence, difficult to capture everything succintly. I think this does capture the main types of documentation and ties back to the principles mentioned in the first sentence.

```suggestion
Effective documentation is essential for any software project to be reproducible, quality assured and transparent. It encompasses both code and the project itself, providing technical details, project management information, records of quality assurance and details of logic, assumptions and decisions made.
```

## What sort of documentation is needed?

> *   **Project-Level:**

Reviewer: **(Question, blocking)**: Does this align with the NHSBSA guidance? If so then we can just link https://nhsbsa.github.io/nhsbsa-digital-playbook/development/dev-documentation/

    *   **README:** Explains the project's purpose, scope, basic setup, usage, and links to other key documentation.
    *   **Installation Guide:** Detailed instructions for getting the project running.
    *   **Usage and Examples:** Clear instructions and examples on how to use the project's features.
    *   **Contribution Guidelines:** How others can contribute (code style, testing, PR process).
    *   **License Information:** Defines legal usage and distribution rights.
    *   **Changelog:** Tracks changes across versions.
    *   **Code of Conduct:** Sets community standards.
    *   **Support Information:** How to get help.
    *   **Design Docs:** High-level overview for complex projects.

>     *   **Accessibility Statement:** If an output is available as a webpage or app, it must have an accessibility statement.

Reviewer 2: Hi @MarkMc1089 just chipping in...
the `dev-documentation` in the DDaT playbook is more about docs that would be checked into source code.

It doesn't cover:
* Project documentation. Often we have multiple source code repositories for a single project. I think the project should use our standard repository for those docs. Our defacto standard is Confluence, but the EAs are looking at this right too now.
* Web application content. Its a good point about the a11y statement, but I think that should be covered in frontend standards. We also need to point out that we should have cookie consent and session timeout UI elements. None of this is in the DDaT playbook, but needs to be.

>     *   **Accessibility Statement:** If an output is available as a webpage or app, it must have an accessibility statement.

Reviewer: Hi both, just to chime in on these points

1. I like Mark's first change linking to the DDaT playbook and also mentioning low/no code solutions
1. On project docs: there might be a need to differentiate between software and analytical projects, especially when we are thinking about RAP - e.g. at silver-level RAP:
    >
    > Code is well-documented including user guidance, explanation of code structure & methodology and [docstrings](https://github.com/NHSDigital/rap-community-of-practice/blob/main/docs/training_resources/python/python-functions.md#documentation) for functions.
    >
    I would say 'methodology' includes decisions / designs around code to produce the analysis. That's not to say that these docs need to go in the repo, but they need to be accessible by anyone who can access the code (i.e. if the code is open then the docs should be too). Often it is just easier to put these docs alongside the code in the same repo. Happy to have this conversation though - also wonder what @MWilson92 thinks.
1. On a11y - can we just include that until it's covered by the frontend standards and then we can just link?

>     *   **Accessibility Statement:** If an output is available as a webpage or app, it must have an accessibility statement.

Reviewer 2: 1. ✅ 

2. Project docs.
I think putting documentation within a codebase is fine if that codebase is the full scope of what's being documented. The problem we have is that we tend to have multiple repos for a single project, and project docs need to span them all. We definitely recommend appropriate comments, and READMEs.

We have an ongoing standard on decision records btw. It would be good to get your input on it. We haven't specified where decision records go, but in your case, you may want to if you want to keep it with the code. Github recommends markdown format ADRs:
* https://adr.github.io/madr/

3. ✅

>     *   **Accessibility Statement:** If an output is available as a webpage or app, it must have an accessibility statement.

Reviewer: Thanks Pat. Created a ticket to track this (#41). Will resolve this discussion

*   **Code-Level:**
    *   **In-Code Comments:** Explanations within the code focusing on the *why* (rationale, complex logic, trade-offs).
    *   **API Documentation:** Standardized comments for functions, classes, modules explaining their purpose, parameters, returns, and usage.
    *   **Detailed Explanations:** Separate documents or wiki pages for complex algorithms, data structures, or specific modules when needed.

> ## Why should we document our code and project?

Reviewer: **(Request, blocking)**: I think we need to call out to the key principles of transparent, reproducible, assured throughout these points. We touch on them implicitly, but this would help tie it back to the larger objectives.

*   **Improves Understanding & Clarity:** Makes code and project purpose comprehensible to others and your future self.
*   **Facilitates Onboarding:** Reduces the time needed for new team members or contributors to become productive.
*   **Supports Maintenance & Debugging:** Makes modifying, fixing and refactoring code safer and easier.
*   **Enables Collaboration & Reuse:** Allows others to effectively use, contribute to, and build upon the work.
*   **Encourages Better Design:** Explaining complex parts often highlights areas for simplification or improvement.
*   **Increases User Adoption:** Makes the project accessible and usable for its intended audience.
*   **Preserves Knowledge:** Prevents critical information loss when team members change.

> *   **Attracts Contributors:** Well-documented projects are more inviting and easier to engage with.

Reviewer: **(Suggestion, blocking)**: I think contribution, collaboration and reuse all come within the same point.

We should document:

* a project's purpose and scope
* how to install or run code or other outputs and their dependencies
* how to use code or other outputs like apps or dashboards
* decisions and changes across a project lifecycle

## Why should we document our code and project?

* **Understanding:** Makes code and project purpose clear, improving [transparency][transparent].
* **Collaboration:** Reduces time to become productive for contributors and other users.
* **Maintenance:** Makes modifying, fixing and refactoring code safer and easier.
* **Improved Design:** Explaining complex parts often highlights areas for simplification or improvement.
* **Accessibility:** Makes the project [accessible][accessible] and usable for its intended audience.
* **Knowledge:** Prevents critical information loss when team members change.
* **Standards:** Provides a record of how a project has been [quality assured][quality-assured] and facilitates [reproducibility][reproducible] and [transparency][transparent].

## How do we create and maintain quality documentation?

> 

Reviewer: **(Suggestion, blocking)**: this looks much better. The final thing is it would benefit from some context to explain why it's broken down under these headings (project-level, etc). Just a sentence will do.

  * **Leverage Tooling**
    * Use linters and formatters to enforce documentation standards.
    * Use tools to generate documentation from code comments.
    * Use templates for consistency.
  * **Maintain Discoverability:** Ensure documentation is easy to find and link related documents together.
  * **Assign Responsibility:** Have clear ownership for creating and maintaining different documentation aspects.

## How do we define success?

* **Understandable:** Documentation enables rapid onboarding of contributors and users.
* **Accurate:** Documentation reflects the state of a project, with documentation updates being part of the regular workflow.
* **Discoverable:** Needed information is easy to locate.
* **Completeness:** Documentation covers all aspects adequately for the intended audience, answering reasonable questions.

## How do we measure success?

* **Reviewers:** [Critical friends][critical-friend] and other [peer-reviewers][peer-review] consider all documentation in addition to code or main outputs.
* **Retrospectives:** Documentation is included in retrospective discussions, and any areas for improvement identified are planned in.

> * **Feedback:** We seek feedback on a project, including its documentation, from contributors and users in general communications and by surveys.

Reviewer: **(Suggestion, blocking)** I think these can be coalesced into a single section 'how do we define success'. We've already outlined the points around understanding, accuracy etc above, so don't need too much detail on those here.

  * **Detailed Explanations:** Separate documents or wiki pages for complex algorithms, data structures, or specific modules when needed.

* **General**
  * **Start Early & Integrate:** Treat documentation as an integral part of the development process, and build updates into your workflow.
  * **Write Clearly and Concisely:** Know your audience for a particular piece of documentation and use straightforward language at the right level.
  * **Focus on the "Why":** Especially for in-code comments, explain the reasoning behind non-obvious decisions.
  * **Use Standard Formats & Locations:** Employ conventional filenames, formats, and place documentation where users expect it.
  * **Leverage Tooling**
    * Use linters and formatters to enforce documentation standards.
    * Use tools to generate documentation from code comments.
    * Use templates for consistency.
  * **Maintain Discoverability:** Ensure documentation is easy to find and link related documents together.
  * **Assign Responsibility:** Have clear ownership for creating and maintaining different documentation aspects.

> ## How do we define success?
> 
> * **Understandable:** Documentation enables rapid onboarding of contributors and users.
> * **Accurate:** Documentation reflects the state of a project, with documentation updates being part of the regular workflow.
> * **Discoverable:** Needed information is easy to locate.
> * **Completeness:** Documentation covers all aspects adequately for the intended audience, answering reasonable questions.
> 
> ## How do we measure success?
> 
> * **Reviewers:** [Critical friends][critical-friend] and other [peer-reviewers][peer-review] consider all documentation in addition to code or main outputs.
> * **Retrospectives:** Documentation is included in retrospective discussions, and any areas for improvement identified are planned in.
> * **Feedback:** We seek feedback on a project, including its documentation, from contributors and users in general communications and by surveys.

Author: ```suggestion
## How do we define success?

* Documentation enables rapid onboarding of contributors and users.
* Documentation reflects the state of a project, with documentation updates being part of the regular workflow.
* Needed information is easy to locate.
* Documentation covers all aspects adequately for the intended audience, answering reasonable questions.

**Lookout for:**

* [Critical friends][critical-friend] and other [peer-reviewers][peer-review]  should consider all documentation in addition to code or main outputs.
* Include documentation in retrospective discussions.
* Seek feedback on project documentation.
```

Reviewer: Looks great. Suggesting a slight tweak 
```
**Lookout for:**
* [Critical friends][critical-friend] and other [peer-reviewers][peer-review] should consider and feedback on documentation, in addition to code and outputs.
* Creating and updating documentation is often overlooked during project planning: be sure to allocate sufficient time for documentation in your estimates.
* Include documentation in retrospective discussions to ensure it's improving and evolving in line with the rest of the analysis.
```
